pkg/operator/oidcdiscoveryendpoint/controller.go

```diff
@@ -0,0 +1,413 @@
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Informer{Informer: informer}, &handler.EnqueueRequestForObject{}, p)
+	if err != nil {
+		return err
+	}
+
+	mgr.Add(&oidcDiscoveryEndpointController{reconciler: r, cache: cache, logger: logger})
+
+	return nil
+}
+
+// openshift-kube-apiserver/bound-sa-token-signing-certs[service-account-xxx.pub]
+func isServiceAccountTokenSigner(meta metav1.Object) bool {
+	// all managed resources are named pod-identity-webhook
+	if meta.GetName() == tokenSignerConfigMapName {
+		return true
+	}
+	return false
+}
+
+type s3EndpointReconciler struct {
+	kubeclientset      *kubernetes.Clientset
+	configclientset    *configset.Clientset
+	logger             log.FieldLogger
+	eventRecorder      events.Recorder
+	infrastructureName string
+	region             string
+}
+
+var _ reconcile.Reconciler = &s3EndpointReconciler{}
+
+func (r *s3EndpointReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	r.logger.Info("reconciling AWS S3 OIDC discovery endpoint on detected change to service account token signer")
+
+	skip, err := r.reconcileClusterAuthentication()
+	if err != nil {
+		r.logger.Error(err)
+		return reconcile.Result{Requeue: true}, err
+	}
+	if skip {
+		return reconcile.Result{}, nil
+	}
+
+	err = r.reconcileS3Endpoint()
+	if err != nil {
+		r.logger.Error(err)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *s3EndpointReconciler) reconcileClusterAuthentication() (bool, error) {
+	r.logger.Debugf("reconciling cluster Authentication serviceAccountIssuer")
+
+	auth, err := r.configclientset.ConfigV1().Authentications().Get(context.TODO(), "cluster", metav1.GetOptions{})
+	if err != nil {
+		return false, err
+	}
+
+	serviceAccountIssuer := r.getIssuerURL()
+	if auth.Spec.ServiceAccountIssuer != "" {
+		r.logger.Debug("Authentication ServiceAccountIssuer already specified")
+		if auth.Spec.ServiceAccountIssuer != serviceAccountIssuer {
+			r.logger.Debug("Authentication ServiceAccountIssuer is not set to the s3 bucket location, skipping OIDC reconciliation")
+			return true, nil
+		}
+		return false, nil
+	}
+
+	auth.Spec.ServiceAccountIssuer = serviceAccountIssuer
+	_, err = r.configclientset.ConfigV1().Authentications().Update(context.TODO(), auth, metav1.UpdateOptions{})
+	return false, err
+}
+
+func (r *s3EndpointReconciler) getIssuerURL() string {
+	return fmt.Sprintf("https://s3-%s.amazonaws.com/%s", r.region, r.getBucketName())
+}
+
+func (r *s3EndpointReconciler) getBucketName() string {
+	return fmt.Sprintf("%s-oidc", r.infrastructureName)
+}
+
+func (r *s3EndpointReconciler) reconcileS3Endpoint() error {
+	r.logger.Debugf("reconciling S3 endpoint resources")
+
+	// Get the root secret and create an AWS client
+	secret, err := r.kubeclientset.CoreV1().Secrets(constants.CloudCredSecretNamespace).Get(context.TODO(), awsannotator.AWSCloudCredSecretName, metav1.GetOptions{})
```

Comment:
Tangential discussion but we are in the process of documenting and supporting a mode of operation where the admin cloud creds can be removed from the cluster after install, and only recreated before an upgrade, leaving the creds to coast in the meantime. (and cco has a r/o cred to monitor and report problems)

Will this mode of operation work with the sts tokens?

Reply:
currently, i'm using those root creds to reconcile to the s3 bucket.  if we don't have those we will either need to 1) mint limited s3 creds for this controller to use or 2) disable ourselves and print a message saying the OIDC endpoint is not available

Reply:
@sjenning I assume these proposed "limited s3 creds" would be permanent Secret-based creds?